Source/WebCore/html/HTMLAttachmentElement.cpp

```diff
@@ -147,27 +158,56 @@ void HTMLAttachmentElement::ensureModernShadowTree(ShadowRoot& root)
     m_innerLegacyAttachment->m_icon = WTFMove(m_icon);
     m_innerLegacyAttachment->m_iconSize = m_iconSize;
     m_innerLegacyAttachment->setIdAttribute(attachmentPreviewIdentifier());
-    container->appendChild(*m_innerLegacyAttachment);
+    m_elementContainer->appendChild(*m_innerLegacyAttachment);
 
     m_elementWithAction = HTMLDivElement::create(document());
     m_elementWithAction->setIdAttribute(attachmentActionIdentifier());
     if (const auto& action = attachmentActionForDisplay(); !action.isEmpty())
         m_elementWithAction->setInnerText(String { action });
-    container->appendChild(*m_elementWithAction);
+    m_elementContainer->appendChild(*m_elementWithAction);
 
     m_elementWithTitle = HTMLDivElement::create(document());
     m_elementWithTitle->setIdAttribute(attachmentTitleIdentifier());
     if (auto title = attachmentTitleForDisplay(); !title.isEmpty())
         m_elementWithTitle->setInnerText(WTFMove(title));
-    container->appendChild(*m_elementWithTitle);
+    m_elementContainer->appendChild(*m_elementWithTitle);
 
     m_elementWithSubtitle = HTMLDivElement::create(document());
     m_elementWithSubtitle->setIdAttribute(attachmentSubtitleIdentifier());
     if (auto subtitle = attachmentSubtitleForDisplay(); !subtitle.isEmpty())
         m_elementWithSubtitle->setInnerText(WTFMove(subtitle));
-    container->appendChild(*m_elementWithSubtitle);
+    m_elementContainer->appendChild(*m_elementWithSubtitle);
+
+    updateSave(attributeWithoutSynchronization(saveQualifiedName()));
 }
 
+void HTMLAttachmentElement::updateSave(const AtomString& eventTypeName)
+{
+    if (!m_elementContainer)
+        return;
+
+    if (eventTypeName.isEmpty() || eventTypeName.contains("'"_s)) {
+        if (m_elementWithSave) {
+            m_elementContainer->removeChild(*m_elementWithSave);
+            m_elementWithSave = nullptr;
+        }
+        return;
+    }
+
+    if (!m_elementWithSave) {
+        m_elementWithSave = HTMLDivElement::create(document());
+        m_elementWithSave->setIdAttribute(attachmentSaveIdentifier());
+        m_elementContainer->appendChild(*m_elementWithSave);
+    }
+
+    StringBuilder saveDispatch;
+    saveDispatch.append("dispatchEvent(new CustomEvent('"_s);
+    saveDispatch.append(eventTypeName);
+    saveDispatch.append("', { bubbles: true, composed: true }))"_s);
+    m_elementWithSave->setAttributeWithoutSynchronization(HTMLNames::onclickAttr, saveDispatch.toAtomString());
```

Comment:
This seems error prone and prone to security vulnerabilities as well.

Can create a new button class that has an `handleEvent` function that calls `dispatchEvent` from C++? 

`PDFDocument.cpp` might have some good examples.

Reply:
`SpinButtonElement` is another good example.

Reply:
Updated with custom listener, thanks for the help.